src/Core/Vipr/FileWriter.cs

```diff
@@ -1,33 +1,89 @@
+        private static ConcurrentDictionary<string, AsyncLock> lockDictionary = new ConcurrentDictionary<string, AsyncLock>();
+        internal static Logger Logger => LogManager.GetLogger("FileWriter");
+
+        public static void WriteAsync(IEnumerable<TextFile> textFilesToWrite, string outputDirectoryPath = null)
         {
             if (!string.IsNullOrWhiteSpace(outputDirectoryPath) && !Directory.Exists(outputDirectoryPath))
                 Directory.CreateDirectory(outputDirectoryPath);
 
-            foreach (var file in textFilesToWrite)
-            {
-                var filePath = file.RelativePath;
-
-                if (!string.IsNullOrWhiteSpace(outputDirectoryPath))
+            /* Write asynchronously in batches of 50
+            * This prevents I/O from slowing down
+            * from an overload of requests
+            */
+            var batchSize = 50;
+            var batchCount = 0;
+            List<Task> tasks = new List<Task>();
+
+            foreach (var file in textFilesToWrite)
+            {
+                tasks = new List<Task>();
+                var filePath = file.RelativePath;
+
+                if (!string.IsNullOrWhiteSpace(outputDirectoryPath))
                     filePath = Path.Combine(outputDirectoryPath, filePath);
 
                 if (!String.IsNullOrWhiteSpace(Environment.CurrentDirectory) &&
                     !Path.IsPathRooted(filePath))
                     filePath = Path.Combine(Environment.CurrentDirectory, filePath);
 
-                if (!Directory.Exists(Path.GetDirectoryName(filePath)))
-                    Directory.CreateDirectory(Path.GetDirectoryName(filePath));
-
-                File.WriteAllText(filePath, file.Contents);
+                if (!Directory.Exists(Path.GetDirectoryName(filePath))) {
+                    try
+                    {
+                        Directory.CreateDirectory(Path.GetDirectoryName(filePath));
+                    } catch (IOException e)
+                    {
+                        Logger.Error("Failed to create directory for file", e);
+                    }
+                }
+
+                // Create a new async task to write the file
+                tasks.Add(WriteToDisk(filePath, file.Contents));
+                batchCount++;
+
+                // Reached the batch limit. Wait for all tasks to complete
+                if (batchCount > batchSize)
+                {
+                    Task t = Task.WhenAll(tasks);
+                    t.Wait();
+                    batchCount = 0;
+                }
+            }
+
+            // Final task wait
+            if (tasks.Count > 0)
+            {
+                Task t = Task.WhenAll(tasks);
+                t.Wait();
             }
         }
+
+        public static async Task WriteToDisk(string filePath, string output)
+        {
+            var lockItem = lockDictionary.GetOrAdd(filePath, new AsyncLock());

```

Comment:
Looks like GetorAdd is not Thread Safe. https://docs.microsoft.com/en-us/dotnet/standard/collections/thread-safe/how-to-add-and-remove-items

One Option would to use 'Lazy' ConcurrentDictionary<string, Lazy < AsyncLock > >.  
https://aka.ms/qb74cz